### ISSUE TITLE
`option_if_let_else` incorrect suggestion, ignores semantics of existing code

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#7991]
I tried this code:

```rust
enum Enum {
    One(u8),
    Two,
}

fn main() {
    let x = Some(Enum::One(1));

    let y = if let Some(one @ Enum::One(..)) = x {
        1
    } else {
        2
    };
}
```

Expected: option_if_let_else should probably not suggest anything here, it can't be trivially replaced with a map or unwrap function.

Instead, clippy suggested replacing the if let with:

```rust
x.map_or(2, |one| 1)
```

This is an incorrect suggestion that will compile but will change the behaviour of the code.

### Meta

**Rust version (`rustc -Vv`):**

```
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0

```

I usually disable `option_if_let_else` because I find it usually makes the code harder to read, but every now and then I'll run pedantic/nursery to see if there are lints I should be enabling. Unlike most of the broken suggestions for this lint, this one will actually compile but break the semantics of the existing code.

@rustbot label +I-suggestion-causes-error +I-false-positive
